src/libslic3r/GCode/SpiralVase.cpp

```diff
@@ -70,28 +138,69 @@ std::string SpiralVase::process_layer(const std::string &gcode)
                 float dist_XY = line.dist_XY(reader);
                 if (dist_XY > 0) {
                     // horizontal move
-                    if (line.extruding(reader)) {
+                    if (line.extruding(reader)) { // We need this to exclude retract and wipe moves!
                         len += dist_XY;
-                        line.set(reader, Z, z + len * layer_height_factor);
-                        if (transition && line.has(E))
-                            // Transition layer, modulate the amount of extrusion from zero to the final value.
-                            line.set(reader, E, line.value(E) * len / total_layer_length);
+                        float factor = len / total_layer_length;
+                        if (transition_in)
+                            // Transition layer, interpolate the amount of extrusion from zero to the final value.
+                            line.set(reader, E, line.e() * factor);
+                        else if (transition_out) {
+                            // We want the last layer to ramp down extrusion, but without changing z height!
+                            // So clone the line before we mess with its Z and duplicate it into a new layer that ramps down E
+                            // We add this new layer at the very end
+                            GCodeReader::GCodeLine transitionLine(line);
+                            transitionLine.set(reader, E, line.e() * (1 - factor));
+                            transition_gcode += transitionLine.raw() + '\n';
+                        }
+                        // This line is the core of Spiral Vase mode, ramp up the Z smoothly
+                        line.set(reader, Z, z + factor * layer_height);
+                        if (smooth_spiral) {
+                            // Now we also need to try to interpolate X and Y
+                            SpiralPoint p(line.x(), line.y()); // Get current x/y coordinates
+                            current_layer->push_back(p);       // Store that point for later use on the next layer
+
+                            if (previous_layer != NULL) {
+                                bool        found    = false;
+                                float       dist     = 0;
+                                SpiralPoint nearestp = nearest_point_on_polygon(p, previous_layer, found, dist);
```

Comment:
Just thinking out loud here.
Not sure if we should add a normal check for the nearest point.
Theoretically such edge cases exists that, but I can't think of a real world model to test it.

![image](https://github.com/SoftFever/OrcaSlicer/assets/103989404/ced45b21-b47f-4240-8349-673ec9889c8a)

Reply:
That's a good point. I think we need to come up with a real world example to justify writing the fix given how complex this is to deal with. Remember that smoothing occurs to the closest line, so you'd need this shape to exist on layer n and nothing similar on layer n-1 for the smoothing to go wrong.

Reply:
https://www.printables.com/model/432767-the-octahedroflake-a-higher-dimensional-analog-of-
This is probably the most closed model I can think of.

Reply:
> https://www.printables.com/model/432767-the-octahedroflake-a-higher-dimensional-analog-of- This is probably the most closed model I can think of.

Yup, models like this are exactly why I made Smooth Spiral configurable on/off configurable. I would argue that models like this shouldn't even be printed in vase mode at all. They rely so heavily on one layer doing something then the next layer doing something completely different. That's not what vase mode is for.